### PR TITLE
1894: map, privates, corporations, setup

### PR DESCRIPTION
**Pull request by Galatolol**
Initial PR for my prototype (1894: https://www.boardgamegeek.com/geeklist/50723/item/8626841#item8626841). Not ready yet, but apparently it is preferred to submit PRs frequently (per #18xxgamesdev).

The map, privates, corporations and setup are done.

### CODE REVIEW

File: lib/engine/game/g_1894/game.rb [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+            Engine::Step::Dividend,
+            Engine::Step::DiscardTrain,
+            G1894::Step::BuyTrain,
+            [Engine::Step::BuyCompany, { blocks: true }],
+          ], round_num: round_num)
+        end
+
+        def setup
+          @late_corporations, @corporations = @corporations.partition do |c|
+            %w[F1 F2 B1 B2].include? c.id
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:272]: use parens

File: lib/engine/game/g_1894/game.rb [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+          remove_extra_trains
+          remove_extra_late_corporations
+
+          @ferry_marker_ability =
+            Engine::Ability::Description.new(type: 'description', description: 'Ferry marker')
+          block_england
+
+          plm = corporations.find { |c| c.id == 'PLM' }
+          paris_tiles_names = %w[X1 X4 X5 X7 X8]
+          paris_tiles = @all_tiles.filter { |t| paris_tiles_names.include? t.name }
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:285]: don't use filter, use parens

File: lib/engine/game/g_1894/game.rb [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+        def revenue_for(route, stops)
+          revenue = super
+          revenue += pc_bonus(route.corporation, stops)
+          revenue += est_le_sud_bonus(route.corporation, stops)
+          revenue
+        end
+
+        def pc_bonus(corp, stops)
+          corp.assigned?('PC') && stops.map(&:hex).find { |hex| hex.assigned?('PC') } ? 10 : 0
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:325]: steps.any? since you don't use hex

File: lib/engine/game/g_1894/game.rb [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        end
+
+        def est_le_sud_bonus(corp, stops)
+          corp.id == 'Est' && stops.map(&:hex).find { |hex| hex.id == 'I2' } ? 20 : 0
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:329]: same

File: lib/engine/game/g_1894/game.rb [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        end
+
+        def ferry_marker_available?
+          hex_by_id(ENGLAND_FERRY_SUPPLY).tile.icons.any? { |icon| icon.name == FERRY_MARKER_ICON }
+        end
+
+        def ferry_marker?(entity)
+          return false unless entity.corporation?
+
+          ferry_markers(entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:339]: !empty?

File: lib/engine/game/g_1894/game.rb [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+
+        def buy_ferry_marker(entity)
+          return unless can_buy_ferry_marker?(entity)
+
+          entity.spend(FERRY_MARKER_COST, @bank)
+          entity.add_ability(@ferry_marker_ability.dup)
+          @log << "#{entity.name} buys a ferry marker for $#{FERRY_MARKER_COST}"
+
+          tile_icons = hex_by_id(ENGLAND_FERRY_SUPPLY).tile.icons
+          tile_icons.delete_at(tile_icons.find_index { |icon| icon.name == FERRY_MARKER_ICON })
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:367]: index

Comment (tobymao) [reply lib/engine/game/g_1894/game.rb:367]: just do tile_icons.reject!...

File: lib/engine/game/g_1894/game.rb [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+
+          graph.clear
+        end
+
+        def block_england
+          england = hex_by_id(ENGLAND_HEX).tile.cities.first
+
+          england.instance_variable_set(:@game, self)
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:375]: what's going on here...

File: lib/engine/game/g_1894/game.rb [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+
+          to_remove = @depot.trains.reverse.find { |t| t.name == '5' }
+          @depot.forget_train(to_remove)
+          @log << "Removing #{to_remove.name} train"
+        end
+
+        def remove_extra_late_corporations
+          return unless @players.size == 3
+
+          to_remove = @late_corporations.filter { |c| %w[F2 B2].include? c.id }
```
Comment (tobymao) [inline lib/engine/game/g_1894/game.rb:395]: fix this

File: lib/engine/game/g_1894/round/stock.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+require_relative '../../../round/stock'
+
+module Engine
+  module Game
+    module G1894
+      module Round
+        class Stock < Engine::Round::Stock
+          protected
+
+          def sold_out_stock_movement(corp)
```
Comment (tobymao) [inline lib/engine/game/g_1894/round/stock.rb:12]: are you just copying 1828? or is there a difference

File: lib/engine/game/g_1894/step/route.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            super
+            return if @game.loading
+
+            return unless route_includes_england?(action.routes) && !@game.ferry_marker?(action.entity)
+
+            raise GameError, 'Cannot run to England without a ferry marker'
+          end
+
+          def route_includes_england?(routes)
+            routes.flat_map(&:connection_hexes).flatten.include?(Engine::Game::G1894::Game::ENGLAND_HEX)
```
Comment (tobymao) [inline lib/engine/game/g_1894/step/route.rb:20]: why flat_map flatten?

File: lib/engine/game/g_1894/step/special_buy.rb [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+module Engine
+  module Game
+    module G1894
+      module Step
+        class SpecialBuy < Engine::Step::SpecialBuy
+          attr_reader :ferry_marker
+
+          def buyable_items(entity)
+            return [@ferry_marker] if @game.loading || @game.can_buy_ferry_marker?(entity)
```
Comment (tobymao) [inline lib/engine/game/g_1894/step/special_buy.rb:13]: what's going on here

File: lib/engine/game/g_1894/step/track.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module Engine
+  module Game
+    module G1894
+      module Step
+        class Track < Engine::Step::Track
+          def legal_tile_rotation?(_entity, hex, tile)
+            if hex.id == @game.class::PARIS_HEX && _hex.tile.color == :green
```
Comment (tobymao) [inline lib/engine/game/g_1894/step/track.rb:9]: what is _hex

File: lib/engine/game/g_1894/step/track.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+              return true if tile.rotation == hex.tile.rotation
+            else
+              super
+            end
+          end
+
+          def update_token!(_action, _entity, tile, old_tile)
+            return unless old_tile.id == @game.class::PARIS_HEX && old_tile.paths.empty?
```
Comment (tobymao) [inline lib/engine/game/g_1894/step/track.rb:17]: don't use unless with multiple statements

File: lib/engine/game/g_1894/step/special_buy.rb (hunk 11 above)
Comment (Galatolol) [reply lib/engine/game/g_1894/step/special_buy.rb:13]: It's the same thing as coal marker in 1828. Here you give the possibility to buy it if the corporation is connected to England.

File: lib/engine/game/g_1894/round/stock.rb (hunk 9 above)
Comment (Galatolol) [reply lib/engine/game/g_1894/round/stock.rb:12]: Sold out is the same. I changed the sold_out? method though.

File: lib/engine/game/g_1894/game.rb (hunk 7 above)
Comment (Galatolol) [reply lib/engine/game/g_1894/game.rb:375]: It's the same as block_va_coalfields in 1828. Here it checks whether to offer the possibility to buy a ferry token (the corporation has to be connected to England).

**Review by tobymao — APPROVED** [review]

**Pull request merged**